### ISSUE TITLE
Selecting multiple objects in the Transform Controls

### ISSUE THREAD

**Issue by Ni55aN** [mrdoob/three.js#6336]
I have a suggestion for the modification Tranform Controls for transformation on multiple objects simultaneously. Should we expect a this in the near future?

**Comment by ShantanuSriv** [comment]
Hi, I have a feature which needs a similar implementation. Has the above request been handled? I don't see the same reflect in the TransformControls but has that been handled in some 3rd-party library which extends TransformControls? If so, can you please refer me to the same?

**Comment by Mugen87** [comment]
@arodic I've seen multiple question in the past where users try to transform a group of objects. What is your recommended way of doing this? Let's say you select a bunch of objects like in this example and then want to transform it:

https://threejs.org/examples/misc_boxselection.html

**Comment by Robyk62** [comment]
> @arodic I've seen multiple question in the past where users try to transform a group of objects. What is your recommended way of doing this? Let's say you select a bunch of objects like in this example and then want to transform it:
> 
> https://threejs.org/examples/misc_boxselection.html

Position the transform control at the center of group bounding box.
Translate each children of the same amount.
Rotate each children around the center of bounding box.
Scale each child the same amount.

i.e. exactely as per transforming meshes inside an object...?

**Comment by arodic** [comment]
Hey @Robyk62, this is a feature that I've developed to a certain degree in the past but never got to integrate it into three.js

The approach that worked for me was to create some sort of SelectionTransformHelper. It was an empty Object3D that I would add to the helper scene and it would act as a middleman between TransformControls and a list of objects (multi-selection). TransformControls would attach to the SelectionTransformHelper, and the helper would have a custom updateMatrixWorld logic that would extract the position/rotation/scale before and after matrix update and then I would apply those transformations to selection.

This is how I would extract transformations from SelectionTransformHelper:

```javascript
updateMatrixWorld(force) {
 	// Extract tranformations before and after matrix update.
	this.matrixWorld.decompose(posOld, quatOld, scaleOld);
	super.updateMatrixWorld(force);
	this.matrixWorld.decompose(pos, quat, scale);

	// Now you can calculate transformation deltas and apply them to a list of objects
}
```

However, the complete solution is far from simple. There are many other problems. For example:

* Handling world vs local transformation
* Handling world vs local transformation in complex hierarchies
* Handling world vs local transformation in complex hierarchies where some of the selected objects are parents/ancestors of others.
* Figuring out where to place/initialize selection helper depending on world/local mode

**Comment by Gllloria** [comment]
Hi! I've been working on this the last days and this solves some of the abovementioned issues (local and world matrix transformations for adding selected objects to new group and back). Hope this helps someone :). What is still missing is where to place the control, and to reposition it every time a new object comes in and the "center of gravity" of the group changes. I'm thankful for Suggestions for this :)
```
var control = new THREE.TransformControls(camera, renderer.domElement);
var tempMatrix = new THREE.Matrix4();
//tempgroup: temporary group for grouping
var tempgroup = new THREE.Group();
scene.add(tempgroup);
tempgroup.userData.selected = [];
tempgroup.userData.prevParent = [];

//For grouping e.g. click on objects
object.addEventListener('click', onGrouping);
//for deselection call onGroupingEnd()

function onGrouping(evt) {
    tempgroup.matrixWorldNeedsUpdate = true;
    if (evt.detail.intersection.object !== undefined) {
        var intersectedObject = evt.detail.intersection.object;

        //Return if object was already selected before
        if (tempgroup.userData.selected.includes(intersectedObject)) {
            return;
        }

        //Recalculate local transform (pos, rot, scale) of intersectedObject for switching parent (tempgroup = new parent)
        intersectedObject.matrixWorldNeedsUpdate = true;
        tempMatrix.getInverse(tempgroup.matrixWorld);
        var intersectedObject_matrix_new = intersectedObject.matrixWorld.premultiply(tempMatrix);
        intersectedObject_matrix_new.decompose(intersectedObject.position, intersectedObject.quaternion, intersectedObject.scale);

        tempgroup.userData.selected.push(intersectedObject);
        tempgroup.userData.prevParent.push(intersectedObject.parent);
        tempgroup.add(intersectedObject);
        // Coloring intersectedObject blue when selecting
        intersectedObject.material.emissive.b = 1;

        if (tempgroup.userData.selected.length > 1) {
            //Attach control to tempgroup
            control.attach(tempgroup);
            //Here I adjust the control size based on boundingSpheres and scales with control.setSize( ... )
        } else {
            //Attach control to intersectedObject
            control.attach(intersectedObject);
            //Here I adjust the control size based on boundingSpheres and scales with control.setSize( ... )
        }

    }
}

//Deselecting
function onGroupingEnd() {
    if (tempgroup.userData.selected !== []) {
        var intersectedObject;
        for (var i = 0; i < tempgroup.userData.selected.length; i++) {
            intersectedObject = tempgroup.userData.selected[i];

            //Recalculate local transform (pos, rot, scale) of intersectedObject for switching back to old parent (prevParent = previous parent)
            intersectedObject.matrixWorldNeedsUpdate = true;
            tempgroup.userData.prevParent[i].matrixWorldNeedsUpdate = true;
            tempMatrix.getInverse(tempgroup.userData.prevParent[i].matrixWorld);
            var intersectedObject_matrix_old = intersectedObject.matrixWorld.premultiply(tempMatrix);
            intersectedObject_matrix_old.decompose(intersectedObject.position, intersectedObject.quaternion, intersectedObject.scale);

            tempgroup.userData.prevParent[i].add(intersectedObject);
            // Coloring intersectedObject normal when deselecting
            intersectedObject.material.emissive.b = 0;
        }
        tempgroup.userData.selected = [];
        tempgroup.userData.prevParent = [];
    }
}
```

**Comment by h-iL** [comment]
> 
> 
> Hi! I've been working on this the last days and this solves some of the abovementioned issues (local and world matrix transformations for adding selected objects to new group and back). Hope this helps someone :). What is still missing is where to place the control, and to reposition it every time a new object comes in and the "center of gravity" of the group changes. `new THREE.Box3().setFromObject(tempgroup)` seems to be the solution for this but I'm thankful for Suggestions for this :)
> 
> ```js
> var control = new THREE.TransformControls(camera, renderer.domElement);
> var tempMatrix = new THREE.Matrix4();
> //tempgroup: temporary group for grouping
> var tempgroup = new THREE.Group();
> scene.add(tempgroup);
> tempgroup.userData.selected = [];
> tempgroup.userData.prevParent = [];
> 
> //For grouping e.g. click on objects
> object.addEventListener('click', onGrouping);
> //for deselection call onGroupingEnd()
> 
> function onGrouping(evt) {
>     tempgroup.matrixWorldNeedsUpdate = true;
>     if (evt.detail.intersection.object !== undefined) {
>         var intersectedObject = evt.detail.intersection.object;
> 
>         //Return if object was already selected before
>         if (tempgroup.userData.selected.includes(intersectedObject)) {
>             return;
>         }
> 
>         //Recalculate local transform (pos, rot, scale) of intersectedObject for switching parent (tempgroup = new parent)
>         intersectedObject.matrixWorldNeedsUpdate = true;
>         tempMatrix.getInverse(tempgroup.matrixWorld);
>         var intersectedObject_matrix_new = intersectedObject.matrixWorld.premultiply(tempMatrix);
>         intersectedObject_matrix_new.decompose(intersectedObject.position, intersectedObject.quaternion, intersectedObject.scale);
> 
>         tempgroup.userData.selected.push(intersectedObject);
>         tempgroup.userData.prevParent.push(intersectedObject.parent);
>         tempgroup.add(intersectedObject);
>         // Coloring intersectedObject blue when selecting
>         intersectedObject.material.emissive.b = 1;
> 
>         if (tempgroup.userData.selected.length > 1) {
>             //Attach control to tempgroup
>             control.attach(tempgroup);
>             //Here I adjust the control size based on boundingSpheres and scales with control.setSize( ... )
>         } else {
>             //Attach control to intersectedObject
>             control.attach(intersectedObject);
>             //Here I adjust the control size based on boundingSpheres and scales with control.setSize( ... )
>         }
> 
>     }
> }
> 
> //Deselecting
> function onGroupingEnd() {
>     if (tempgroup.userData.selected !== []) {
>         var intersectedObject;
>         for (var i = 0; i < tempgroup.userData.selected.length; i++) {
>             intersectedObject = tempgroup.userData.selected[i];
> 
>             //Recalculate local transform (pos, rot, scale) of intersectedObject for switching back to old parent (prevParent = previous parent)
>             intersectedObject.matrixWorldNeedsUpdate = true;
>             tempgroup.userData.prevParent[i].matrixWorldNeedsUpdate = true;
>             tempMatrix.getInverse(tempgroup.userData.prevParent[i].matrixWorld);
>             var intersectedObject_matrix_old = intersectedObject.matrixWorld.premultiply(tempMatrix);
>             intersectedObject_matrix_old.decompose(intersectedObject.position, intersectedObject.quaternion, intersectedObject.scale);
> 
>             tempgroup.userData.prevParent[i].add(intersectedObject);
>             // Coloring intersectedObject normal when deselecting
>             intersectedObject.material.emissive.b = 0;
>         }
>         tempgroup.userData.selected = [];
>         tempgroup.userData.prevParent = [];
>     }
>     control.detach();
> }
> ```

Hi @Gllloria, may i ask what does the 'evt' stands for? I tried run this function  while 'dragging-changed' but i get error message 'event.detail is undefined'

Here s my code

`
console.log('yeah it is running')

import { TransformControls } from './js/TransformControls.js'
import { OrbitControls } from "https://cdn.jsdelivr.net/npm/three@0.119.1/examples/jsm/controls/OrbitControls.min.js"

import { SelectionBox } from './js/SelectionBox.js'
import { SelectionHelper } from './js/SelectionHelper.js'
import { MeshLambertMaterial } from './js/three.module.js'


var allSelected=[]

////////////////////////////////////////////////////////////////////////////////

const scene=new THREE.Scene()

const camera = new THREE.PerspectiveCamera(75, window.innerWidth / window.innerHeight, 0.1, 1000)

const renderer = new THREE.WebGLRenderer()

renderer.setSize(window.innerWidth, window.innerHeight)
document.body.appendChild(renderer.domElement)


////////////////////////////////////////////////////////////////////////////////

const axesHelper = new THREE.AxesHelper(5)

scene.add(axesHelper)

////////////////////////////////////////////////////////////////////////////////

const backgroundTexture = new THREE.CubeTextureLoader().load(['./textures/cube/clouds/3.png', './textures/cube/clouds/1.png', './textures/cube/clouds/top.png', './textures/cube/clouds/white.png', './textures/cube/clouds/2.png', './textures/cube/clouds/4.png'])
scene.background = backgroundTexture


////////////////////////////////////////////////////////////////////////////////



const light = new THREE.SpotLight(0xffffff, 1.5)
light.position.set(0, 500, 2000)
light.angle=Math.PI/9

light.castShadow = true
scene.add(light)



////////////////////////////////////////////////////////////////////////////////



const tempGroup = new THREE.Group()
scene.add(tempGroup)

tempGroup.userData.selected = []
tempGroup.userData.prevParent = []


const geometry = new THREE.BoxGeometry(2, 2, 2)

for (let i = 0; i < 20; i++)
{
    const object = new THREE.Mesh(geometry, new THREE.MeshLambertMaterial({ color: Math.random() * 0xffffff }))

    object.position.x = Math.random() * 50 - 25
    object.position.y = Math.random() * 50 - 25
    object.position.z = Math.random() * 50 - 25

    scene.add(object)

}


////////////////////////////////////////////////////////////////////////////////

let selected = new THREE.Object3D()
scene.add(selected)


const orbitControls = new OrbitControls(camera, renderer.domElement)


const transformControls = new TransformControls(camera, renderer.domElement)

transformControls.setMode('rotate')
scene.add(transformControls)
transformControls.attach(selected)



transformControls.addEventListener('dragging-changed', function (event) {

    orbitControls.enabled = !event.value
   onGroupingStart(event)
})


window.addEventListener('keydown', function (event) {

    switch (event.key || event.keyCode) {

        case "g":
            transformControls.setMode('translate')
            break

        case "r":
            transformControls.setMode('rotate')
            break

        case "s":
            transformControls.setMode('scale')
            break

        case "z":
            orbitControls.enabled = false
            break

        case "x":
            orbitControls.enabled = true
            break


    }
})


document.addEventListener('dblclick', function () {

    if (orbitControls.enabled) {
        orbitControls.enabled = false
    }

    else orbitControls.enabled = true
})

////////////////////////////////////////////////////////////////////////////////

const selectionBox = new SelectionBox(camera, scene)
const helper = new SelectionHelper(selectionBox, renderer, 'selectBox')



document.addEventListener('pointerdown', function (event) {


    console.log('pointer down')

    for (const item of selectionBox.collection) {
              
        try {
            if (item.material.type === 'MeshLambertMaterial')
                item.material.emissive.set(0x000000);

        }
        catch (err) {
            console.log(item.material)
        }     

    }

    selectionBox.startPoint.set(
        (event.clientX / window.innerWidth) * 2 - 1,
        -(event.clientY / window.innerHeight) * 2 + 1,
        0.5)

})



document.addEventListener('pointermove', function (event) {

    if (helper.isDown) {
        console.log('pointer move')

       

        for (let i = 0; i < selectionBox.collection.length; i++) {
            
            try {
                if (selectionBox.collection[i].material.type === 'MeshLambertMaterial')
                    selectionBox.collection[i].material.emissive.set(0x000000);

            }
            catch (err) {
                console.log(selectionBox.collection[i].material)
            }

        }

        selectionBox.endPoint.set(
            (event.clientX / window.innerWidth) * 2 - 1,
            -(event.clientY / window.innerHeight) * 2 + 1,
            0.5)

        allSelected = selectionBox.select()




        for (let i = 0; i < allSelected.length; i++) {

            try {
                if (allSelected[i].material.type === 'MeshLambertMaterial')
                    allSelected[i].material.emissive.set(0xffffff);

            }
            catch (err) {
                console.log(allSelected[i].material)
            }
        }
    }

})

document.addEventListener('pointerup', function (event) {

    console.log('pointer up')


    selectionBox.endPoint.set(
        (event.clientX / window.innerWidth) * 2 - 1,
        -(event.clientY / window.innerHeight) * 2 + 1,
        0.5)

    const allSelected = selectionBox.select()


   
    //allSelected.forEach(o => tempGroup.add(o))
    


    for (let i = 0; i < allSelected.length; i++) {

        try {
            if (allSelected[i].material.type === 'MeshLambertMaterial')
                allSelected[i].material.emissive.set(0xffffff);

        }
        catch (err) {
            console.log(allSelected[i].material)
        }
    }
    
})


////////////////////////////////////////////////////////////////////////////////

var tempMatrix = new THREE.Matrix4()


function onGroupingStart(event)
{
    tempGroup.matrixWorldNeedsUpdate = true

    
    if (event.detail.intersection.object !== undefined)
    {        
        var intersectedObject = event.detail.intersection.object

        if (tempGroup.userData.selected.includes(intersectedObject))
        {
            return
        }

        intersectedObject.matrixWorldNeedsUpdate = true
        tempMatrix.getInverse(tempGroup.matrixWorld)

        var intersectedObject_matrix_new = intersectedObject.matrixWorld.premultiply(tempMatrix)
        intersectedObject_matrix_new.decompose(intersectedObject.position, intersectedObject.quaternion, intersectedObject.scale)

        tempGroup.userData.selected.push(intersectedObject)
        tempGroup.userData.prevParent.push(intersectedObject.parent)

        tempGroup.add(intersectedObject)
        intersectedObject.material.emissive.b = 1

        if (tempGroup.userData.selected.length > 1) {
            transformControls.attach(tempGroup)
        }
        else
        {
            transformControls.attach(intersectedObject)
        }



    }

}


function onGroupingEnd() {

    if (tempGroup.userData.selected !== [])
    {
        var intersectedObject

        for (let i = 0; i < tempGroup.userData.selected.length; i++)
        {
            intersectedObject = tempGroup.userData.selected[i]
            intersectedObject.matrixWorldNeedsUpdate = true

            tempGroup.userData.prevParent[i].matrixWorldNeedsUpdate = true

            tempMatrix.getInverse(tempGroup.userData.prevParent[i].matrixWorld)

            let intersectedObject_matrix_old = intersectedObject.matrixWorld.premultiply(tempMatrix)
            intersectedObject_matrix_old.decompose(intersectedObject.position, intersectedObject.quaternion, intersectedObject.scale)

            tempGroup.userData.prevParent[i].add(intersectedObject)
            intersectedObject.material.emissive.b=0

        }

        tempGroup.userData.selected = []
        tempGroup.userData.prevParent = []

    }
    transformControls.detach


}



////////////////////////////////////////////////////////////////////////////////

camera.position.z = 2

window.addEventListener('resize', onWindowResize, false)

function onWindowResize()
{
    camera.aspect = window.innerWidth / window.innerHeight
    camera.updateProjectionMatrix()
    renderer.setSize(window.innerWidth, window.innerHeight)
    render()
}

function render()
{
    renderer.render(scene, camera)
}

var animate = function () {
    requestAnimationFrame(animate)
    render()
}

animate()
window.focus()

`
